File: Java/Hard/WrongAnswer/equal-rational-numbers_2.java
class Solution {
    public boolean isRationalEqual(String s, String t) {
        
        int i = s.indexOf("(");
        int j = t.indexOf("(");
        String s1 = s.substring(0, i);
        String s2 = s.substring(i+1, s.length()-1);
        String t1 = t.substring(0, j);
        String t2 = t.substring(j+1, t.length()-1);
        String s3 = s1 + s2;
        String t3 = t1 + t2;
        String s4 = s1 + s2 + s2 + s2 + s2 + s2 + s2 + s2 + s2 + s2;
        String t4 = t1 + t2 + t2 + t2 + t2 + t2 + t2 + t2 + t2 + t2;
        if (s1.equals(t1) && s2.equals(t2)) {
            return true;
        } else if (s3.equals(t3)) {
            return true;
        } else if (s4.equals(t4)) {
            return true;
        } else {
            return false;
        }
    }
}

File: Python3/Hard/RuntimeError/equal-rational-numbers_2.py
class Solution:
    def isRationalEqual(self, s: str, t: str) -> bool:
        
        def parse(s):
            if '(' not in s:
                return int(s), 1, 0
            i = s.index('(')
            return int(s[:i]), 1, int(s[i + 1:-1])
        def reduce(a, b, c):
            g = math.gcd(a, b)
            a //= g
            b //= g
            g = math.gcd(b, c)
            b //= g
            c //= g
            return a, b, c
        def add(a, b, c, d, e):
            return a * d + b * c, b * d, e
        def eq(a, b, c, d, e):
            a, b, c = reduce(a, b, c)
            d, e, f = reduce(d, e, f)
            return a == d and b == e and c == f
        a, b, c = parse(s)
        d, e, f = parse(t)
        if c == 0 and f == 0:
            return a * e == d * b
        if c == 0:
            return eq(a, b, c, d * b + a, e * b)
        if f == 0:
            return eq(a * e, b * e, c * e, d, e)
        return eq(a * e * (f + 1), b * e * (f + 1), c * e * (f + 1), d * e * (f + 1), e * (f + 1))